script/prerender-workers/work.ts

```diff
@@ -0,0 +1,262 @@
+    this.worker.on('message', async(receiptHandle: string) => {
+      try {
+        // Pass the receiptHandle back to the main loop
+        if (this.resolvePromise) { this.resolvePromise(receiptHandle); }
+
+        // The worker is now idle again
+        pushWorker(this);
+      } catch (e) {
+        // Error on the main thread; Stop the whole container and let someone else retry
+        console.error(e.message, e.stack);
+        process.exit(1);
+      }
+    });
+
+    // This usually happens if there's a connection error
+    this.worker.on('exit', (code: number) => {
+      try {
+        console.log(`Worker thread exited with status code ${code}`);
+
+        if (this.rejectPromise) { this.rejectPromise(new Error(code.toString())); }
+
+        // Start a brand new worker thread if terminated by an error
+        if (code !== 0) { pushWorker(new SQSWorker()); }
+      } catch (e) {
+        // Error on the main thread; Stop the whole container and let someone else retry
+        console.error(e.message, e.stack);
+        process.exit(1);
+      }
+    });
+
+    // These should not happen as we try to catch the errors and exit(1) instead,
+    // but just in case we missed something, we log them
+    this.worker.on('error', console.error);
+    this.worker.on('messageerror', console.error);
+  }
+
+  public async startWork(message: Message) {
+    // This promise will be resolved when the worker is done
+    const promise = new Promise<string>((resolve, reject) => {
+      this.resolvePromise = resolve;
+      this.rejectPromise = reject;
+    });
+
+    // Begin work on a separate thread
+    this.worker.postMessage(message);
+
+    return promise;
+  }
+}
+
+// Idle worker thread (LIFO) queue
+// LIFO is cheaper in complexity and we don't really care about the order of the rendering
+
+const idleWorkers: SQSWorker[] = [];
+const resolveWorkerPromises: Array<(worker: SQSWorker) => void> = [];
+
+function pushWorker(worker: SQSWorker) {
+  // Check if someone is already waiting for a worker
+  const resolveWorkerPromise = resolveWorkerPromises.pop();
+
+  if (resolveWorkerPromise) {
+    // console.log('Main thread already waiting for worker thread; sending it directly');
+
+    resolveWorkerPromise(worker);
+  } else {
+    // console.log('Main thread not ready yet; adding worker thread to idle worker queue');
+
+    idleWorkers.push(worker);
+  }
+}
+
+async function popWorker() {
+  // Check if an idle worker is available
+  const worker = idleWorkers.pop();
+
+  if (worker) {
+    // console.log('Main thread got worker thread from idle worker queue');
+
+    return worker;
+  } else {
+    // console.log('All worker threads busy; main thread waiting for available worker thread');
+
+    // Return a promise that will be resolved when pushWorker is called
+    return new Promise<SQSWorker>((resolve) => { resolveWorkerPromises.push(resolve); });
+  }
+}
+
+async function sqsHeartbeat(messages: Message[]) {
```

Comment:
what is sqs heartbeat?

Reply:
I should probably write this as a comment in there :)